src/loaders.py

```diff
@@ -9,52 +11,80 @@
 from steppy.base import BaseTransformer
 
 from .retinanet import DataEncoder
-from .pipeline_config import MEAN, STD
+from .pipeline_config import MEAN, STD, SEED
+from .utils import get_target_size
 
 
-class RandomSubsetSampler(Sampler):
-    def __init__(self, data_size, sample_size):
-        self.data_size = data_size
-        self.sample_size = sample_size
+class FixedSizeSampler(Sampler):
+    def __init__(self, images_data, *,  sample_size=None, **kwargs):
+        self.images_data = images_data.sample(frac=1, random_state=SEED)
+        self.sample_size = len(self.images_data) if sample_size is None else min(sample_size, len(self.images_data))
 
     def __iter__(self):
-        return iter(torch.randperm(self.data_size)[:self.sample_size].long())
+        indices = random.sample(range(len(self.images_data)), self.sample_size)
+        return iter(indices)
+
+    def __len__(self):
+        return self.sample_size
+
+
+class AspectRatioSampler(Sampler):
+    def __init__(self, images_data, *, batch_size=8, sample_size=None):
+        self.images_data = images_data
+        self.batch_size = batch_size
+        self.sample_size = len(self.images_data) if sample_size is None else min(sample_size, len(self.images_data))
+        self.sample_size = self.sample_size // self.batch_size * self.batch_size
+
+        self.indices = self.images_data.sort_values('aspect_ratio').index.tolist()
+
+    def __iter__(self):
+        subset = sorted(random.sample(range(len(self.indices)), self.sample_size))
+        indices = np.array([self.indices[i] for i in subset])\
+            .reshape(self.sample_size // self.batch_size, self.batch_size)
```

Comment:
@taraspiotr this // was already calculated in the __init__ do we really want to have it here ?

Reply:
@jakubczakon I've changed it a bit